File: algorithms/2_medium/M/03660_jump_game_ix/prefix.c
static inline int max_int(const int a, const int b) {
    return a > b ? a : b;
}

static inline int min_int(const int a, const int b) {
    return a < b ? a : b;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* maxValue(const int* nums, const int len, int* return_len) {
    if (len <= 0 || nums == NULL || return_len == NULL) {
        fprintf(stderr, "invalid input");
        exit(1);
    }

    int* prefix = malloc((size_t)len * sizeof(*prefix));
    if (prefix == NULL) {
        fprintf(stderr, "alloc fail");
        exit(1);
    }
    *return_len = len;
    prefix[0] = nums[0];
    for (int i = 1; i < len; i += 1) {
        prefix[i] = max_int(nums[i], prefix[i - 1]);
    }

    int curr_min = INT_MAX;
    int curr_max = 0;
    for (int i = len - 1; i >= 0; i -= 1) {
        if (prefix[i] > curr_min) {
            prefix[i] = curr_max;
            curr_min = min_int(curr_min, nums[i]);
        } else {
            curr_max = prefix[i];
            curr_min = nums[i];
        }
    }

    return prefix;
}

File: algorithms/2_medium/M/03660_jump_game_ix/prefix.rs
impl Solution {
    pub fn max_value(nums: Vec<i32>) -> Vec<i32> {
        if nums.is_empty() { return Vec::default(); }
        let len = nums.len();
        let mut prefix = vec![0_i32; len];
        prefix[0] = nums[0];
        for i in 1..len {
            prefix[i] = nums[i].max(prefix[i - 1]);
        }
        let mut curr_min = i32::MAX;
        let mut curr_max = 0;
        for i in (0..len).rev() {
            if prefix[i] > curr_min {
                prefix[i] = curr_max;
                curr_min = curr_min.min(nums[i]);
            } else {
                curr_max = prefix[i];
                curr_min = nums[i];
            }
        }

        prefix
    }
}
